metafields.go

```diff
@@ -0,0 +1,196 @@
+	// ProductImageResource is the resource for metafields related to product images.
+	ProductImageResource = "product_image"
+	// ProductVariantResource is the resource for metafields related to product variants.
+	ProductVariantResource = "variants"
+	// ShopResource is the resource for metafields related to the shop.
+	ShopResource = "shop"
+)
+
+// Metafield is a flexible way to attach additional information to a Shopify resource (e.g. Product, Collection, etc.).
+type Metafield struct {
+	// ID is the unique ID of the metafield.
+	ID int64
+	// Description is the description of the information that the metafield contains.
+	Description string
+	// Key is the key of the metafield. Keys can be up to 30 characters long and can contain alphanumeric characters, hyphens, underscores, and periods.
+	Key string
+	// Namespace is the container name for a group of metafields. Grouping metafields within a namespace prevents your metafields from conflicting with other metafields with the same key name. Must have between 3-20 characters.
+	Namespace string
+	// Resource is the unique ID of the resource that the metafield is attached to and the type of resource that the metafield is attached to.
+	Resource MetafieldResource
+	// Value is the data stored in the metafield. The value is always stored as a string, regardless of the metafield's type.
+	Value string
+	// MetafieldType is the type of data that the metafield stores in the `value` field.
+	Type string
+	// CreatedAt is the  date and time (ISO 8601 format) when the metafield was created.
+	CreatedAt time.Time
+	// UpdatedAt is the  date and time (ISO 8601 format) when the metafield was last updated.
+	UpdatedAt time.Time
+}
+
+// MetafieldResource holds the query data for a resource
+type MetafieldResource struct {
+	// OwnerID is the unique ID of the resource that the metafield is attached to.
+	OwnerID int64
+	// OwnerResource is the type of resource that the metafield is attached to.
+	OwnerResource string
+}
+
+// MetafieldRepository maintains the metafields of a shop.
+type MetafieldRepository interface {
+	// List gets all the metafields
+	List(query MetafieldQuery) (Metafields, error)
+}
+
+// MetafieldQuery are properties that can be used to filter the returned metafields
+// See https://shopify.dev/api/admin-rest/2022-07/resources/metafield#get-metafields
+type MetafieldQuery struct {
+	// Resource is the resource and ID that the metafields are attached to
+	Resource MetafieldResource
+}
+
+// ErrInvalidMetafieldType is thrown when an invalid metafield type was found
+type ErrInvalidMetafieldType struct {
+	msg string
+}
+
+// Error returns the specified message
+func (e ErrInvalidMetafieldType) Error() string {
+	return e.msg
+}
+
+// NewErrInvalidMetafieldType creates a new ErrInvalidMetafieldType
+func NewErrInvalidMetafieldType(metafieldType string) ErrInvalidMetafieldType {
+	return ErrInvalidMetafieldType{fmt.Sprintf("The metafield type was invalid for :  %s", metafieldType)}
+}
```

Comment:
Then just call it `ErrInvalidMetafieldType{metaType: ""}`.
```go
type ErrInvalidMetafieldType struct {
	metaType string
}

// Error returns the specified message
func (e ErrInvalidMetafieldType) Error() string {
	return fmt.Sprintf("The metafield type was invalid for: %s", e.metaType)
}
```

Also, what does `%s` do? I'm lazy to look it up.

Reply:
This is not actually needed now because I am defaulting to string. %s is just string.